### PR TITLE
devdocs on julia internal locks

### PR DESCRIPTION
**Pull request by vtjnash**

### CODE REVIEW

File: doc/devdocs/locks.rst [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+    doesn't exist right now
+
+    fix: create it
+
+* flisp
+
+    recursive (through macros), but caller might also be holding locks
+
+    fix: unknown
```
Comment (yuyichao) [inline doc/devdocs/locks.rst]: Flisp should be fine with recursion and doesn't need a lock.

File: doc/devdocs/locks.rst [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    fix: create it
+
+* codegen
+
+    recursive (through static_eval), but caller might also be holding locks (due to staged functions)
+
+    other issues?
+
+    fix: prohibit codegen while holding any other lock?
```
Comment (yuyichao) [inline doc/devdocs/locks.rst]: I guess we can check `ptls->current_task->locks.len != 0`? Calling codegen from a `_NOGC` lock would be a serious error.

File: doc/devdocs/locks.rst [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+
+The following is a leaf lock (level 2), and only acquires level 1 locks (safepoint) internally:
+
+   * symbol_table
+
+The following are level 3 locks, which can only acquire level 1 or level 2 locks internally:
+
+   * dump
+   * typecache
```
Comment (yuyichao) [inline doc/devdocs/locks.rst]: This lock is also currently wrong. It only protect cache lookup and insertion but it doesn't make the lookup-and-construct atomic.

Comment (vtjnash) [inline doc/devdocs/locks.rst]: lookup-and-construct atomicity should be mostly handled by the `Method->writelock` now

File: doc/devdocs/locks.rst [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+       flisp itself is already threadsafe, this lock only protects the ``jl_ast_context_list_t`` pool
+
+
+The following is a leaf lock (level 2), and only acquires level 1 locks (safepoint) internally:
+
+   * symbol_table
+
+The following are level 3 locks, which can only acquire level 1 or level 2 locks internally:
+
+   * dump
```
Comment (yuyichao) [inline doc/devdocs/locks.rst]: Depending on how likely we'll have this running on multiple thread (maybe not if we can directly work on bytecode) we can probably get rid of this one fairly easily by storing the states on stack.

File: doc/devdocs/locks.rst (hunk 3 above)
Comment (yuyichao) [inline doc/devdocs/locks.rst]: Not for `apply_type`?

File: doc/devdocs/locks.rst (hunk 4 above)
Comment (vtjnash) [inline doc/devdocs/locks.rst]: true. we also need to make the usage of the incremental deserialize a toplevel expression lock

File: doc/devdocs/locks.rst [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+Below are all of the locks that exist in the system and the mechanism in use
+that prevents them from being a potential deadlock.
+
+The following are definitely leaf locks (level 1), and must not try to acquire any other lock:
+
+   * safepoint
+
+       note that this lock is acquired implicitly by ``JL_LOCK`` and ``JL_UNLOCK``.
+       use the ``_NOGC`` variants to avoid that for level 1 locks.
```
Comment (yuyichao) [inline doc/devdocs/locks.rst]: I think I had this in the comment, but maybe it's also worth mentioning here that a `_NOGC` lock is not allowed to do any allocation or hit any safepoints. Or in the word you use here, any safepoint including allocation, enable/disable GC, exception frame enter and restore, implicitly acquires the safepoint lock.

File: doc/devdocs/locks.rst [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+   * typeinf
+
+       this one is perhaps one of the most tricky ones, since type-inference can be invoked from many points
+
+The following is the root locks, meaning no other lock shall be held when trying to acquire it:
+
+   * toplevel
+
+       this should be held while attempting a top-level action (such as making a new type or defining a new method):
+       trying to obtain this lock inside a staged function will cause a deadlock condition!
```
Comment (vtjnash) [inline doc/devdocs/locks.rst]: also should note here that it's unclear if any code can safely run in parallel with a toplevel expression

File: doc/devdocs/locks.rst (hunk 3 above)
Comment (vtjnash) [inline doc/devdocs/locks.rst]: oh, oops. though this comment was on `typeinf`. I didn't realize `apply_type` was broken too – will add a comment

File: doc/devdocs/locks.rst [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+The following strategies are used to ensure that the code is dead-lock free:
+
+ 1. structure code such that only one lock will need to be acquired at a time
+
+ 2. always acquire shared locks in the same order, as given by the table below
+
+ 3. avoid constructs that expect to need unrestricted recursion
+
+Below are all of the locks that exist in the system and the mechanism in use
+that prevents them from being a potential deadlock.
```
Comment (tkelman) [inline doc/devdocs/locks.rst]: the locks themselves aren't "a potential deadlock," it's the way in which they're acquired that can be

File: doc/devdocs/locks.rst [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+The following are definitely leaf locks (level 1), and must not try to acquire any other lock:
+
+   * safepoint
+
+       Note that this lock is acquired implicitly by ``JL_LOCK`` and ``JL_UNLOCK``.
+       use the ``_NOGC`` variants to avoid that for level 1 locks.
+
+       While holding this lock, the code must not do any allocation or hit any safepoints.
+       Note that there are safepoint when doing allocation, enabling / disabling GC,
```
Comment (tkelman) [inline doc/devdocs/locks.rst]: safepoints

File: doc/devdocs/locks.rst [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+       entering / restoring exception frame, and taking / releasing locks.
```
Comment (tkelman) [inline doc/devdocs/locks.rst]: exception frames ?

File: doc/devdocs/locks.rst [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+   * dump
+   * typecache
+
+The following is a level 4 locks, which can only recurse to acquire level 1, 2, or 3 locks:
+
+   * Method->writelock
+
+     but note that this is violated by staged functions!
+
+The following are level 5 locks, which can only recurse to acquire to acquire locks at lower levels:
```
Comment (tkelman) [inline doc/devdocs/locks.rst]: to acquire ~~to acquire~~

File: doc/devdocs/locks.rst [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+The following is a leaf lock (level 2), and only acquires level 1 locks (safepoint) internally:
+
+   * symbol_table
+
+The following are level 3 locks, which can only acquire level 1 or level 2 locks internally:
+
+   * dump
+   * typecache
+
+The following is a level 4 locks, which can only recurse to acquire level 1, 2, or 3 locks:
```
Comment (tkelman) [inline doc/devdocs/locks.rst]: "lock," singular

File: doc/devdocs/locks.rst [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+     but note that this is violated by staged functions!
+
+The following are level 5 locks, which can only recurse to acquire to acquire locks at lower levels:
+
+   * MethodTable->writelock
+
+     but note that this is violated by staged functions!
+
+The following is a proposed level 6 locks, which can only recurse to acquire locks at lower levels:
```
Comment (tkelman) [inline doc/devdocs/locks.rst]: "lock," singular

File: doc/devdocs/locks.rst [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+   * staged
+
+       this theoretical lock would create a priority inversion from the `method->writelock` (level 4),
+       but only prohibiting only running any staging function in parallel
```
Comment (tkelman) [inline doc/devdocs/locks.rst]: only prohibiting, or prohibiting only?

File: doc/devdocs/locks.rst [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+
+    not certain of whether there are issues here or what they are. staging functions, of course, are a source of deadlocks here.
+
+    fix: unknown
+
+* staged
+
+    possible solution to prevent staged functions from causing deadlock.
+
+    this theoretic lock would create a priority inversion such that the Method and MethodTable write locks could be released
```
Comment (tkelman) [inline doc/devdocs/locks.rst]: theoretical

File: doc/devdocs/locks.rst [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    by ensuring that no staging functions can run in parallel allow this level 6 lock to protect staged function conflicts (a level 4 operation)
+
+    fix: create it
+
+* typecache
+
+    this only protect cache lookup and insertion but it doesn't make the lookup-and-construct atomic
```
Comment (tkelman) [inline doc/devdocs/locks.rst]: protects

File: doc/devdocs/locks.rst (hunk 13 above)
Comment (vtjnash) [inline doc/devdocs/locks.rst]: `staging` function is correct

Comment (tkelman) [inline doc/devdocs/locks.rst]: what is a staging function then? needs elaboration

Comment (yuyichao) [inline doc/devdocs/locks.rst]: Functions that we are, well, staging.....

Comment (tkelman) [inline doc/devdocs/locks.rst]: what is staging then? is this related to staged functions or not? either way, what is the difference between "staged function" as used elsewhere vs "staging function" here?

Comment (yuyichao) [inline doc/devdocs/locks.rst]: Staged function is what you create with `@generated`. Staging is the state it's in when we are generating it.

Comment (vtjnash) [inline doc/devdocs/locks.rst]: The function that creates a staged function. I realize we rarely have distinguished between the two. A macro produces a macro expansion as a staging function produces a generated function.

Comment (vtjnash) [inline doc/devdocs/locks.rst]: Elsewhere, I've also referred to them as "generated lambda templates", but that's mostly as a derivative of "lambda template". I think `staging` is more clear as the active tense of which `generated` is the past tense (result).

Comment (tkelman) [inline doc/devdocs/locks.rst]: ahk yeah with that written explanation somewhere visible, that does make sense.

File: doc/devdocs/locks.rst (hunk 7 above)
Comment (vtjnash) [inline doc/devdocs/locks.rst]: I don't understand the distinction you are making

Comment (tkelman) [inline doc/devdocs/locks.rst]: "all the locks that exist in the system" are not "a potential deadlock," it's incorrect acquisition or release of them that is a potential deadlock. a lock and its acquisition/release are not the same thing.

say "the mechanism for using them in a way that avoids potential deadlocks"[](url)

File: doc/devdocs/locks.rst [hunk 16]
```diff
@@ -0,0 +1,128 @@
+.. currentmodule:: Base
+
+****************************************************
+Proper maintenance and care of multi-threading locks
+****************************************************
+
+The following strategies are used to ensure that the code is dead-lock free
+(generally by addressing the 4th Coffman condition: circular wait).
```
Comment (tkelman) [inline doc/devdocs/locks.rst]: please add a reference for this

File: doc/devdocs/locks.rst [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+
+The following is a proposed level 5 lock, which can only recurse to acquire locks at lower levels:
+
+   * staged
+
+       this theoretical lock would create a priority inversion from the `method->writelock` (level 3),
+       but only prohibiting running any staging function in parallel
+       (thus allowing temporary release of the MethodTable and Method locks)
+
+The following is a level 6 locks, which can only recurse to acquire locks at lower levels:
```
Comment (tkelman) [inline doc/devdocs/locks.rst]: level 6 lock, singular

File: doc/devdocs/locks.rst [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+
+   * codegen
+
+The following is an almost root locks (level end-1), meaning only the root look may be held when trying to acquire it:
```
Comment (tkelman) [inline doc/devdocs/locks.rst]: almost root lock, singular

File: doc/devdocs/locks.rst [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+
+   * typeinf
+
+       this one is perhaps one of the most tricky ones, since type-inference can be invoked from many points
+
+The following is the root locks, meaning no other lock shall be held when trying to acquire it:
```
Comment (tkelman) [inline doc/devdocs/locks.rst]: root lock, singular

File: doc/devdocs/locks.rst (hunk 16 above)
Comment (vtjnash) [inline doc/devdocs/locks.rst]: this is standard terminology. I'm not adding "refer to your textbook on deadlock avoidance (Coffman, 1971)"

Comment (tkelman) [inline doc/devdocs/locks.rst]: I'd wager that the majority of Julia users, and a fair number of its developers, have never heard of that reference. Title, author, year.

Comment (vtjnash) [inline doc/devdocs/locks.rst]: It's the *only* topic covered in the wikipedia article on deadlock: https://en.wikipedia.org/wiki/Deadlock. It seems not all textbooks give the name, and fewer still provide a reference to the original paper. But I finally found one on the second page of google results:

E. G. Coffman, Jr; M. J. Elphick; A. Shoshani, System Deadlocks, Computing Surveys, Vol. 3, No. 2, June 1971.
http://people.cs.umass.edu/~mcorner/courses/691J/papers/TS/coffman_deadlocks/coffman_deadlocks.pdf

Comment (tkelman) [inline doc/devdocs/locks.rst]: If you don't come from a formal CS background, you're also not likely to have learned these with a standard numbering.

File: doc/devdocs/locks.rst [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+(generally by addressing the 4th Coffman condition: circular wait).
+
+ 1. structure code such that only one lock will need to be acquired at a time
+
+ 2. always acquire shared locks in the same order, as given by the table below
+
+ 3. avoid constructs that expect to need unrestricted recursion
+
+Below are all of the locks that exist in the system and the mechanisms
+for using them the avoids the potential for deadlocks (no Ostrich algorithm allowed here):
```
Comment (tkelman) [inline doc/devdocs/locks.rst]: "mechanism for using them that avoids" or "mechanisms for using them that avoid"

File: doc/devdocs/locks.rst (hunk 16 above)
Comment (vtjnash) [inline doc/devdocs/locks.rst]: yes, that's why we have search engines

Comment (JeffBezanson) [inline doc/devdocs/locks.rst]: I think the policy with devdocs is that anything is better than nothing.

Comment (yuyichao) [inline doc/devdocs/locks.rst]: FWIW, I don't have CS background and I've never heard of these before. However, in this case, the name is clear enough for google to return the right result.

Comment (tkelman) [inline doc/devdocs/locks.rst]: Something is better than nothing, but would rather try to make what's in this repo as approachable as we can. The purpose of them is so people who don't understand things yet can learn more about them.

Comment (vtjnash) [inline doc/devdocs/locks.rst]: The purpose of this document is for people who do understand what they are doing, and have commit rights, to have a reference document against which to validate changes.

Comment (tkelman) [inline doc/devdocs/locks.rst]: that's not really "Documentation of Julia's Internals" then

File: doc/devdocs/locks.rst [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+    not certain of whether there are issues here or what they are. staging functions, of course, are a source of deadlocks here.
+
+    fix: unknown
+
+* staged
+
+    possible solution to prevent staged functions from causing deadlock.
+
+    this theoretical lock would create a priority inversion such that the Method and MethodTable write locks could be released
+    by ensuring that no staging functions can run in parallel allow this level 5 lock to protect staged function conflicts (a level 3 operation)
```
Comment (tkelman) [inline doc/devdocs/locks.rst]: grammar here makes no sense
